keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2469 @@
+# between ResourceClass and ResourceClaim
+# parameters.
+initCommand:
+- /usr/local/bin/acme-gpu-init
+- --cluster
+- my-cluster
+---
+apiVersion: core.k8s.io/v1alpha1
+kind: ResourceClass
+metadata:
+  name: acme-gpu
+driverName: gpu.example.com
+parameters:
+  apiVersion: gpu.example.com/v1
+  kind: GPUInit
+  name: acme-gpu-init
+```
+
+#### Partial GPU allocation
+
+As a user, I want to use a GPU as accelerator, but don't need exclusive access
+to that GPU. Running my workload with just 2Gb of memory is sufficient. This is
+supported by the ACME GPU hardware. I know that the administrator has created
+an "acme-gpu" ResourceClass.
+
+For a simple trial, I create a Pod directly where two containers share the same subset
+of the GPU:
+```
+apiVersion: gpu.example.com/v1
+kind: GPURequirements
+metadata:
+  name: device-consumer-gpu-parameters
+memory: "2Gi"
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: device-consumer
+spec:
+  resourceClaims:
+  - name: "gpu" # this name gets referenced below under "claims"
+    template:
+      resourceClassName: "acme-gpu"
+      parameters:
+        apiVersion: gpu.example.com/v1
+        kind: GPURequirements
+        name: device-consumer-gpu-parameters
+  containers:
+  - name: workload
+    image: my-app
+    command: ["/bin/program"]
+    resources:
+      requests:
+        memory: "64Mi"
+        cpu: "250m"
+      limits:
+        memory: "128Mi"
+        cpu: "500m"
+      claims:
+        - "gpu"
+  - name: monitor
+    image: my-app
+    command: ["/bin/other-program"]
+    resources:
+      requests:
+        memory: "32Mi"
+        cpu: "25m"
+      limits:
+        memory: "64Mi"
+        cpu: "50m"
+      claims:
+      - "gpu"
+```
+
+This request triggers resource allocation on a node that has a GPU device with
+2Gi of memory available and then the Pod runs on that node. The remaining
+capacity of the GPU may be usable for other pods, with constrains like alignment
+to segment sizes ensured by the resource driver.
+The lifecycle of the resource
+allocation is tied to the lifecycle of the Pod.
+
+In production, a similar PodTemplateSpec in a Deployment will be used.
+
+#### Network-attached accelerator
+
+As a data center operator, I want to place accelerators in a separate enclosure
+where they can get connected to different compute nodes through [Compute
+Express Link™ (CXL™)](https://www.computeexpresslink.org/). This is more
+cost-effective because not all workloads need these accelerators and more
+flexible because worker nodes can be added or removed independently from
+accelerators.
+
+Jobs then compete for these accelerators by requesting them through a
+ResourceClaim. The kube-scheduler ensures that all required resources for a pod
+are reserved for it (CPU, RAM, and the ResourceClaim) before scheduling the pod
+onto a worker node. The resource driver then dynamically attaches the
+accelerator and detaches it again when the pod completes.
```

Comment:
Now this is in, all my earlier quibbles about user stories and central control are addressed.

It's not a problem to add the IP video cam user story - but it's also 100% fine by me to merge this KEP without that.

Reply:
Let's leave it out then. More text just means more things to review before the deadline...